redash/serializers/__init__.py

```diff
@@ -25,21 +25,22 @@ def public_widget(widget):
         'created_at': widget.created_at
     }
 
-    if widget.visualization and widget.visualization.id:
-        query_data = models.QueryResult.query.get(widget.visualization.query_rel.latest_query_data_id).to_dict()
+    v = widget.visualization
+    if v and v.id:
         res['visualization'] = {
-            'type': widget.visualization.type,
-            'name': widget.visualization.name,
-            'description': widget.visualization.description,
-            'options': json_loads(widget.visualization.options),
-            'updated_at': widget.visualization.updated_at,
-            'created_at': widget.visualization.created_at,
+            'type': v.type,
+            'name': v.name,
+            'description': v.description,
+            'options': json_loads(v.options),
+            'updated_at': v.updated_at,
+            'created_at': v.created_at,
             'query': {
+                'id': v.query_rel.id,
                 'query': ' ',  # workaround, as otherwise the query data won't be loaded.
-                'name': widget.visualization.query_rel.name,
-                'description': widget.visualization.query_rel.description,
-                'options': {},
-                'latest_query_data': query_data
+                'name': v.query_rel.name,
+                'data_source_id': v.query_rel.data_source_id,
```

Comment:
Is this really needed?

Reply:
You've [asked this before](https://github.com/getredash/redash/pull/3659#discussion_r275154351). Are you implying that you've read my answer and think it _isn't_ needed?

Reply:
I had a feeling of dejavu, but somehow lost track of the previous time I asked about it. 🤦‍♂ 

Anyway, it feels wrong to expose this just because of how the code works. The only reason we make this check there is for cases when the query has no data source associated with it:

https://github.com/getredash/redash/blob/92391e3cbc3fedac3380a44c57c9d8addeeb3238/client/app/services/query.js#L520-L524

We can move this logic to the backend and return the appropriate error there in such cases, making this `if` unnecessary.

Reply:
17fa5e0